#**Car Price Predector**

DataSet is downloaded from kaggle

**Importing libraries**


In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

**Getting data**

In [ ]:
df = pd.read_csv('car data.csv')
df.head()

Car_Name  Year  Selling_Price  ...  Seller_Type  Transmission Owner
0     ritz  2014           3.35  ...       Dealer        Manual     0
1      sx4  2013           4.75  ...       Dealer        Manual     0
2     ciaz  2017           7.25  ...       Dealer        Manual     0
3  wagon r  2011           2.85  ...       Dealer        Manual     0
4    swift  2014           4.60  ...       Dealer        Manual     0

[5 rows x 9 columns]

**Checking missing values**


In [ ]:
df.isnull().sum()

Car_Name         0
Year             0
Selling_Price    0
Present_Price    0
Kms_Driven       0
Fuel_Type        0
Seller_Type      0
Transmission     0
Owner            0
dtype: int64

In this data set there is no null values

-----------------------------------------

**Getting categorical and numerical variables**

In [ ]:
cat_var = []
num_var = []
for feature in df.columns:
    if df[feature].dtype == 'O':
        cat_var.append(feature)
    else:
        num_var.append(feature)

print(cat_var)
print('Length' , len(cat_var))
print('--------------------------')
print(num_var)
print('Length' , len(num_var))
print('--------------------------')

['Car_Name', 'Fuel_Type', 'Seller_Type', 'Transmission']
Length 4
--------------------------
['Year', 'Selling_Price', 'Present_Price', 'Kms_Driven', 'Owner']
Length 5
--------------------------


In [ ]:
for feature in cat_var:
    print(feature ," " , len(df[feature].unique()))

Car_Name   98
Fuel_Type   3
Seller_Type   2
Transmission   2


From above we sen that there are 98 unique variables in Car_name column ,So droping Car_name column


**Droping car_name**

In [ ]:
df.drop('Car_Name', axis = 1 , inplace = True)

In [ ]:
df.head()

Year  Selling_Price  Present_Price  ...  Seller_Type Transmission Owner
0  2014           3.35           5.59  ...       Dealer       Manual     0
1  2013           4.75           9.54  ...       Dealer       Manual     0
2  2017           7.25           9.85  ...       Dealer       Manual     0
3  2011           2.85           4.15  ...       Dealer       Manual     0
4  2014           4.60           6.87  ...       Dealer       Manual     0

[5 rows x 8 columns]

**Treansforming Year to Years**

In [ ]:
df['Years'] = 2021 - df['Year']
df.drop('Year' , axis  = 1 , inplace = True)

In [ ]:
df.head()

Selling_Price  Present_Price  Kms_Driven  ... Transmission Owner Years
0           3.35           5.59       27000  ...       Manual     0     7
1           4.75           9.54       43000  ...       Manual     0     8
2           7.25           9.85        6900  ...       Manual     0     4
3           2.85           4.15        5200  ...       Manual     0    10
4           4.60           6.87       42450  ...       Manual     0     7

[5 rows x 8 columns]

In [ ]:
 from sklearn.preprocessing import OneHotEncoder,LabelEncoder

In [ ]:
cat_var = [feature for feature in df.columns if df[feature].dtype == 'O']
cat_var

['Fuel_Type', 'Seller_Type', 'Transmission']

In [ ]:
labelEncoder = LabelEncoder()

In [ ]:
df['Fuel_Type']

0      Petrol
1      Diesel
2      Petrol
3      Petrol
4      Diesel
        ...  
296    Diesel
297    Petrol
298    Petrol
299    Diesel
300    Petrol
Name: Fuel_Type, Length: 301, dtype: object

In [ ]:
df['Fuel_Type'] = labelEncoder.fit_transform(df['Fuel_Type'])
df.head()

Selling_Price  Present_Price  Kms_Driven  ...  Transmission Owner Years
0           3.35           5.59       27000  ...        Manual     0     7
1           4.75           9.54       43000  ...        Manual     0     8
2           7.25           9.85        6900  ...        Manual     0     4
3           2.85           4.15        5200  ...        Manual     0    10
4           4.60           6.87       42450  ...        Manual     0     7

[5 rows x 8 columns]

In [ ]:
df = pd.get_dummies(df, drop_first=True)
df.head()

Selling_Price  Present_Price  ...  Seller_Type_Individual  Transmission_Manual
0           3.35           5.59  ...                       0                    1
1           4.75           9.54  ...                       0                    1
2           7.25           9.85  ...                       0                    1
3           2.85           4.15  ...                       0                    1
4           4.60           6.87  ...                       0                    1

[5 rows x 8 columns]

In [ ]:
from sklearn.feature_selection import mutual_info_regression

In [ ]:
from sklearn.model_selection import train_test_split

Splitting dependent and independent variable


In [ ]:
X = df.drop('Selling_Price' ,axis = 1)
Y = df['Selling_Price']

Train test split

In [ ]:
x_train,x_test,y_train,y_test = train_test_split(X,Y , test_size=0.2, random_state=0)

#**Feature Selection**

Feature imporetance

In [ ]:
mutual_info = mutual_info_regression(X,Y)

In [ ]:
mutual_data = pd.Series(mutual_info , index= X.columns )
important_feature_weight = mutual_data.sort_values(ascending = False)

In [ ]:
important_feature_weight

Present_Price             1.294932
Seller_Type_Individual    0.535973
Years                     0.255548
Kms_Driven                0.162154
Fuel_Type                 0.157923
Transmission_Manual       0.056343
Owner                     0.033452
dtype: float64

We can drop owner and Transmission_Manual       

In [ ]:
df.head()

Selling_Price  Present_Price  ...  Seller_Type_Individual  Transmission_Manual
0           3.35           5.59  ...                       0                    1
1           4.75           9.54  ...                       0                    1
2           7.25           9.85  ...                       0                    1
3           2.85           4.15  ...                       0                    1
4           4.60           6.87  ...                       0                    1

[5 rows x 8 columns]

In [ ]:
X = df.drop('Selling_Price' ,axis = 1)
Y = df.iloc[: , 0]

In [ ]:
x_train,x_test,y_train,y_test = train_test_split(X.values,Y.values , test_size=0.2, random_state=0)

In [ ]:
print(x_train.shape ,"," , x_test.shape , ',' , y_train.shape, ',' , y_test.shape)

(240, 7) , (61, 7) , (240,) , (61,)


In [ ]:
df.shape

(301, 8)

In [ ]:
!pip install optuna

     |████████████████████████████████| 308 kB 5.8 MB/s 
     |████████████████████████████████| 209 kB 62.2 MB/s 
     |████████████████████████████████| 80 kB 6.7 MB/s 
     |████████████████████████████████| 75 kB 4.0 MB/s 
     |████████████████████████████████| 149 kB 43.2 MB/s 
     |████████████████████████████████| 49 kB 5.3 MB/s 
     |████████████████████████████████| 112 kB 53.6 MB/s 
  Created wheel for pyperclip: filename=pyperclip-1.8.2-py3-none-any.whl size=11136 sha256=d166a2bc70accc47811d5d452b6bb34a4a74bd24a92849a9868ee57a64a7d5b6
  Stored in directory: /root/.cache/pip/wheels/9f/18/84/8f69f8b08169c7bae2dde6bd7daf0c19fca8c8e500ee620a28
Successfully built pyperclip


In [ ]:
import optuna
import sklearn.svm
from sklearn.ensemble import RandomForestRegressor
from xgboost import XGBRegressor


#**Hyper parameter tunning optimization**

**Optuna**

In [ ]:
def get_best_model():

    def RF_objective(RF_trial):
    # Regressor = trial.suggest_categorical('Regressor', ['RandomForest', 'XGBoost'])   
    # if Regressor == 'RandomForest':
        RF_params = {
            'n_estimators'      : RF_trial.suggest_int('n_estimators', 200, 2000,10),
            'max_features'      : RF_trial.suggest_categorical('max_features' , ["auto", "sqrt","log2"]),
            'max_depth'         : int(RF_trial.suggest_float('max_depth', 10, 100, log=True)),
            'min_samples_split' : RF_trial.suggest_int('min_samples_split' ,2,14,1),
            'min_samples_leaf'  : RF_trial.suggest_int('min_samples_leaf' , 1,10,1),
            'criterion'         : RF_trial.suggest_categorical('criterion',["squared_error", "absolute_error", "poisson"]),
        }

        RF_Reg = sklearn.ensemble.RandomForestRegressor(**RF_params)

        return sklearn.model_selection.cross_val_score(RF_Reg, x_train, y_train, n_jobs=-1, cv=3).mean()

    

    def XGB_objectives(XGB_trial):
        XGB_params = {
            'n_estimators'     : XGB_trial.suggest_int('n_estimators', 0, 1000,10),
            'max_depth'        : XGB_trial.suggest_int('max_depth', 2, 25),
            'reg_alpha'        : XGB_trial.suggest_int('reg_alpha', 0, 5),
            'reg_lambda'       : XGB_trial.suggest_int('reg_lambda', 0, 5),
            'min_child_weight' : XGB_trial.suggest_int('min_child_weight', 0, 5),
            'gamma'            : XGB_trial.suggest_int('gamma', 0, 5),
            'learning_rate'    : XGB_trial.suggest_loguniform('learning_rate',0.005,0.5),
            'colsample_bytree' : XGB_trial.suggest_discrete_uniform('colsample_bytree',0.1,1,0.01),
            'nthread'          : -1
        }
        XGB_Reg = XGBRegressor(**XGB_params)
        
        return sklearn.model_selection.cross_val_score(XGB_Reg, x_train, y_train, n_jobs=-1, cv=3).mean()


    study = optuna.create_study(direction='maximize')
    study.optimize(RF_objective, n_trials=100)
    RF_trial = study.best_trial
   
    study = optuna.create_study(direction='maximize')
    study.optimize(XGB_objectives, n_trials=100)
    XGB_trial = study.best_trial


    print('Random forest Regressor')
    print('Accuracy: {}'.format(RF_trial .value))
    print("Best hyperparameters: {}".format(RF_trial .params))
    print("-------------")
    print('XGB Regressor')
    print('Accuracy: {}'.format(XGB_trial.value))
    print("Best hyperparameters: {}".format(XGB_trial.params))



In [ ]:
get_best_model()

[I 2021-12-11 05:24:35,350] A new study created in memory with name: no-name-0373a3dc-dd72-43aa-a5c4-decb693ccfbe
[I 2021-12-11 05:24:38,448] Trial 0 finished with value: 0.7230345091885456 and parameters: {'n_estimators': 440, 'max_features': 'auto', 'max_depth': 79.05553055484796, 'min_samples_split': 6, 'min_samples_leaf': 7, 'criterion': 'poisson'}. Best is trial 0 with value: 0.7230345091885456.
[I 2021-12-11 05:24:40,470] Trial 1 finished with value: 0.7826726688027287 and parameters: {'n_estimators': 540, 'max_features': 'sqrt', 'max_depth': 47.72633699707709, 'min_samples_split': 4, 'min_samples_leaf': 3, 'criterion': 'squared_error'}. Best is trial 1 with value: 0.7826726688027287.
[I 2021-12-11 05:24:44,126] Trial 2 finished with value: 0.5702367915050234 and parameters: {'n_estimators': 960, 'max_features': 'log2', 'max_depth': 73.77221700700133, 'min_samples_split': 4, 'min_samples_leaf': 10, 'criterion': 'poisson'}. Best is trial 1 with value: 0.7826726688027287.
[I 2021-1

Random forest Regressor
Accuracy: 0.8972698456965222
Best hyperparameters: {'n_estimators': 990, 'max_features': 'auto', 'max_depth': 29.871752784324265, 'min_samples_split': 2, 'min_samples_leaf': 1, 'criterion': 'squared_error'}
-------------
XGB Regressor
Accuracy: 0.9218090679050458
Best hyperparameters: {'n_estimators': 780, 'max_depth': 19, 'reg_alpha': 0, 'reg_lambda': 1, 'min_child_weight': 2, 'gamma': 0, 'learning_rate': 0.17138021923098956, 'colsample_bytree': 0.84}




*   Random forest Regressor
Accuracy: 0.8972698456965222

*   XGB Regressor
Accuracy: 0.9218090679050458



**Genetic Algorithm**

In [ ]:
!pip install tpot

     |████████████████████████████████| 87 kB 2.9 MB/s 
     |████████████████████████████████| 173.5 MB 10 kB/s 
     |████████████████████████████████| 160 kB 47.7 MB/s 
  Created wheel for stopit: filename=stopit-1.1.2-py3-none-any.whl size=11952 sha256=3159be28ca7f2f7aac213835e10693ade7288483b7e9a65a5ed5573b8416a73d
  Stored in directory: /root/.cache/pip/wheels/e2/d2/79/eaf81edb391e27c87f51b8ef901ecc85a5363dc96b8b8d71e3
Successfully built stopit
  Attempting uninstall: xgboost
    Found existing installation: xgboost 0.90
    Uninstalling xgboost-0.90:
      Successfully uninstalled xgboost-0.90


In [ ]:
from tpot import TPOTRegressor

tpot = TPOTRegressor(generations=5, population_size=50, verbosity=2 , offspring_size= 12)
tpot.fit(x_train, y_train)
# tpot.export('tpot_boston_pipeline.py')

Optimization Progress:   0%|          | 0/110 [00:00<?, ?pipeline/s]


Generation 1 - Current best internal CV score: -4.0809891912839

Generation 2 - Current best internal CV score: -2.4409799335434967

Generation 3 - Current best internal CV score: -2.4409799335434967

Generation 4 - Current best internal CV score: -2.4409799335434967

Generation 5 - Current best internal CV score: -2.4409799335434967

Best pipeline: XGBRegressor(input_matrix, learning_rate=1.0, max_depth=4, min_child_weight=1, n_estimators=100, n_jobs=1, objective=reg:squarederror, subsample=0.7500000000000001, verbosity=0)


TPOTRegressor(generations=5, offspring_size=12, population_size=50, verbosity=2)

In [ ]:
from sklearn.model_selection import ShuffleSplit
cv = ShuffleSplit(n_splits=5 , random_state= 0 , test_size=0.2)

In [ ]:
tpot.score(x_train, y_train)

-3.397566205432582e-05

In [ ]:
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import ShuffleSplit

In [ ]:
XGBoost = XGBRegressor( learning_rate=0.1, max_depth=10, min_child_weight=2, n_estimators=100, n_jobs=1, objective='reg:squarederror', subsample=0.8500000000000001, verbosity=0)
XGBoost.fit(x_train , y_train)
XGBoost.score(x_test , y_test)


0.9675022101022176



*   XGBoost Regressor : Accuracy -> 0.9675022101022176

Get Accuracy of 96%




In [ ]:
y_pred=XGBoost.predict(x_test)


In [ ]:
import pickle

In [ ]:
with open('RF_Price_Predection_model2.pkl' , 'wb') as file:
    pickle.dump(XGBoost,file)